### PR TITLE
Return consistent `Meter` for a given `MeterProvider`

### PR DESCRIPTION
**Pull request by jtescher**
## Changes

Adds a meter cache to return consistent `Meter` instances for the same scope. Meters have their own instrument caches so this avoids double instrument registration.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by shaun-cox** [comment]
@jtescher, I would kindly request we accept or reject #1328 before merging this, as otherwise I'll have merge conflicts to fix in #1328 I suspect... (and I've been waiting a day or so for final approval.) Thanks.

File: opentelemetry-sdk/src/metrics/meter_provider.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/metrics/meter_provider.rs:132]: Given the fact that obtaining `meter` is not free, we should mention in the documentation that, meter (and instruments) should be obtained once and cached by user to avoid perf hit. Not blocking this PR, but wanted to call it out here.

Comment (jtescher) [reply opentelemetry-sdk/src/metrics/meter_provider.rs:132]: I think we can do some benchmarking in a follow up and have better guidelines on metric perf in general. The use case for accessing a meter is different than a tracer or logger as instruments are more likely to be a fixed set. And even in those cases repeatedly creating/accessing instruments could be the place that needs perf considerations.

**Review by shaun-cox — APPROVED** [review]

**Review by lalitb — APPROVED** [review]

**Pull request merged**
